clients/go/cmd/zbctl/internal/commands/root.go

```diff
@@ -170,19 +174,80 @@ func parseAddress() (address string, port string) {
 	address = DefaultAddressHost
 	port = DefaultAddressPort
 
-	if len(addressFlag) > 0 {
-		address = addressFlag
-	} else if addressEnv, exists := os.LookupEnv(zbc.GatewayAddressEnvVar); exists {
-		address = addressEnv
-	}
+	lenOfHostFlag := len(hostFlag)
+	lenOfPortFlag := len(portFlag)
+	hostEnv, hostEnvExists := os.LookupEnv(zbc.GatewayHostEnvVar)
+	portEnv, portEnvExists := os.LookupEnv(zbc.GatewayPortEnvVar)
+	addressEnv, addressEnvExists := os.LookupEnv(zbc.GatewayAddressEnvVar)
+
+	ConditionChain{
+		validConditionFunc: func() bool {
+			return lenOfHostFlag > 0 || lenOfPortFlag > 0
+		},
+		execFunc: func() {
+			address = getValueByFlagAndOrByEnvOrDefaultValue(hostFlag, zbc.GatewayHostEnvVar, DefaultAddressHost)
+			port = getValueByFlagAndOrByEnvOrDefaultValue(portFlag, zbc.GatewayPortEnvVar, DefaultAddressPort)
+		},
+		elseBranch: &ConditionChain{
+			validConditionFunc: func() bool {
+				return len(addressFlag) > 0
+			},
+			execFunc: func() {
+				address, port = parseHostAndPortFromAddress(addressFlag, DefaultAddressHost, DefaultAddressPort)
+			},
+			elseBranch: &ConditionChain{
+				validConditionFunc: func() bool {
+					return hostEnvExists
+				},
+				execFunc: func() {
+					address = hostEnv
+					port = getValueByFlagAndOrByEnvOrDefaultValue(portFlag, zbc.GatewayPortEnvVar, DefaultAddressPort)
+				},
+				elseBranch: &ConditionChain{
+					validConditionFunc: func() bool {
+						return portEnvExists
+					},
+					execFunc: func() {
+						port = portEnv
+						address = getValueByFlagAndOrByEnvOrDefaultValue(hostFlag, zbc.GatewayHostEnvVar, DefaultAddressHost)
+					},
+					elseBranch: &ConditionChain{
+						validConditionFunc: func() bool {
+							return addressEnvExists
+						},
+						execFunc: func() {
+							address, port = parseHostAndPortFromAddress(addressEnv, DefaultAddressHost, DefaultAddressPort)
+						},
+					},
+				},
+			},
+		},
+	}.run()
+
+	return
+}
 
-	if strings.Contains(address, ":") {
-		parts := strings.Split(address, ":")
+func parseHostAndPortFromAddress(parsedAddress string, defaultAddress string, defaultPort string) (string, string) {
+	var address = defaultAddress
+	var port = defaultPort
+	if strings.Contains(parsedAddress, ":") {
+		parts := strings.Split(parsedAddress, ":")
 		address = parts[0]
 		port = parts[1]
 	}
+	return address, port
+}
+
+func getValueByFlagAndOrByEnvOrDefaultValue(flagValue string, envName string, defaultValue string) string {
```

Comment:
Instead of getting all of these from outside and passing them in, we could encapsulate all of it in the function. So either we built the address from host and port or we'd call `getFullAddress` which would get the relevant things (defaults, env name, etc) and decide what to use. I think that would help shortening the other code and it would nicely encapsulate all of that logic. It would also then not need such a long and descriptive name imo